modules/junit-jupiter/src/main/java/org/testcontainers/junit/jupiter/TestcontainersExtension.java

```diff
@@ -0,0 +1,76 @@
+package org.testcontainers.junit.jupiter;
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.platform.commons.util.Preconditions;
+import org.testcontainers.lifecycle.Startable;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+
+class TestcontainersExtension implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback {
+
+    @Override
+    public void beforeAll(final ExtensionContext context) throws IllegalAccessException {
```

Comment:
You could reduce code duplication between `beforeAll` and `afterAll` by having a method `forEachSharedContainer(Consumer<Field>)` that does the same thing as this method, but instead of `startContainer` calls `Consumer::consume` with the field.